Make BIGNUM code available in the FIPS provider

mattcaswell opened the pull request:
This PR makes the necessary changes to allow BIGNUM to be used from inside the FIPS provider. The BIGNUM RAND code is temporarily disabled for FIPS until the RAND code is also available (see #9035). If the RAND PR gets approved first then we can drop the RAND disabling commit from this PR. Otherwise it can be removed by some later PR.

slontis commented on crypto/bn/bn_ctx.c:
```diff
@@ -86,8 +86,11 @@ struct bignum_ctx {
+    OPENSSL_CTX *libctx;
 };
 
+#ifndef FIPS_MODE
```
Where is the #endif associated with this #ifndef ?

mattcaswell replied:
I try to do that too for longer spans, but forgot in this instance. Added now.

slontis commented on crypto/bn/bn_print.c:
```diff
@@ -286,6 +289,7 @@ int BN_asc2bn(BIGNUM **bn, const char *a)
     return 1;
 }
 
+#ifndef FIPS_MODE
```
It could be pulled in if needed by making BIO_snprintf something that gets passed to the provider.

mattcaswell replied:
I'm hoping we won't need any of these functions.

levitte commented on crypto/bn/build.info:
```diff
@@ -7,6 +7,17 @@ SOURCE[../../libcrypto]=\
+        {- $target{bn_asm_src} -} \
+        bn_recp.c bn_mont.c bn_mpi.c bn_exp2.c bn_gf2m.c bn_nist.c \
+        bn_const.c bn_x931p.c bn_intern.c bn_dh.c \
+        bn_rsa_fips186_4.c
```
Aren't you getting tired of the repetitions?  (I've seen that quite a number of times now)

Personally, I use perl templating, like this:

```
{- our @src = ( qw( bn_add.c bn_div.c bn_exp.c bn_lib.c bn_ctx.c bn_mul.c bn_mod.c
                    bn_print.c bn_rand.c bn_shift.c bn_word.c bn_blind.c
                    bn_kron.c bn_sqrt.c bn_gcd.c bn_prime.c bn_sqr.c
                    bn_recp.c bn_mont.c bn_mpi.c bn_exp2.c bn_gf2m.c bn_nist.c
                    bn_const.c bn_x931p.c bn_intern.c bn_dh.c
                    bn_rsa_fips186_4.c ), $target{bn_asm_src} ); -}
LIBS=../../libcrypto
SOURCE[../../libcrypto]={- join(' ', @src) -} bn_err.c bn_depr.c bn_srp.c
SOURCE[../../providers/fips]={- join(' ', @src) -}
```

But maybe this is a good reason to have local variables in build.info files...

```
$COMMONSRC=\
        bn_add.c bn_div.c bn_exp.c bn_lib.c bn_ctx.c bn_mul.c bn_mod.c \
        bn_print.c bn_rand.c bn_shift.c bn_word.c bn_blind.c \
        bn_kron.c bn_sqrt.c bn_gcd.c bn_prime.c bn_sqr.c \
        {- $target{bn_asm_src} -} \
        bn_recp.c bn_mont.c bn_mpi.c bn_exp2.c bn_gf2m.c bn_nist.c \
        bn_const.c bn_x931p.c bn_intern.c bn_dh.c \
        bn_rsa_fips186_4.c
LIBS=../../libcrypto
SOURCE[../../libcrypto]=$COMMONSRC bn_err.c bn_depr.c bn_srp.c
SOURCE[../../providers/fips]=$COMMONSRC
```

Would that be helpful?

mattcaswell replied:
I went with your suggestion. I'm not sure whether having local variables in the build.info files might be more than is needed.

slontis replied:
Depends on how much work is involved - If it is  easy to add, it certainly looks a lot cleaner.

levitte replied:
It shouldn't be too hard to add.  It's a fairly simple string replace scheme.

mattcaswell replied:
In that case...sure. Why not?

levitte commented on crypto/bn/bn_print.c:
```diff
@@ -286,6 +289,7 @@ int BN_asc2bn(BIGNUM **bn, const char *a)
     return 1;
 }
```
If we want to be pure about it, all the BN_xx2bn and BN_bn2xx functions should be moved to a different unit.  They have nothing to do with printing per se, even though printing may use them.

bn_string.c, maybe?

NOTE: not necessarily subject for this PR...

levitte replied:
(it would allow excluding bn_print.c from being built into the FIPS module, and is one less `#ifndef FIPS_MODE`)

mattcaswell replied:
I created a bn_conv.c file and moved all these functions into it. That of course makes this PR look bigger than it is - its just moving code from one place to another.

mattcaswell commented on crypto/bn/bn_ctx.c:
```diff
@@ -116,35 +119,48 @@ static void ctxdbg(BIO *channel, const char *text, BN_CTX *ctx)
     } OSSL_TRACE_END(BN_CTX)
+#else
+# define CTXDBG(str, ctx) do {}  while(0)
+#endif
```
Done

mattcaswell commented:
I pushed an update addressing all feedback so far. I also added a new commit to handle some assembler changes which I forgot about. I'm hoping that will fix the travis failure...but I couldn't actually recreate it.

mattcaswell commented:
Unfortunately the CIs did not agree (make doc-nits failure). Fixup commit pushed.

Please reconfirm?

premamotion commented:
I got it, thanks!   
  
  
  

  
  

  
  
>   
> On Jun 11, 2019 at 7:16 PM,  <Richard Levitte (mailto:notifications@github.com)>  wrote:
>   
>   
>   
>
> @levitte  approved this pull request.
>
>   
>
> —
>  You are receiving this because you are subscribed to this thread.
>  Reply to this email directly,   view it on GitHub (https://github.com/openssl/openssl/pull/9130?email_source=notifications&email_token=AAGGCPKJPTZOOMKAUGTKELLPZ7FWBA5CNFSM4HW5FMP2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB3GJIVY#pullrequestreview-248288343), or   mute the thread (https://github.com/notifications/unsubscribe-auth/AAGGCPIBVW56A7PMVPFN3T3PZ7FWBANCNFSM4HW5FMPQ).
>
>   
>

mattcaswell commented:
Pushed. Thanks.
